# 线性回归（初步）

**快速提示**

1. 经济学中的回归，核心目标通常不是停留在相关关系，而是讨论一个因素对另一个因素的**因果影响**：$x$ 是否会影响 $y$？会如何影响 $y$？
2. 回归系数是回答因果问题的重要工具。加入控制变量以后，斜率有了“其他变量保持不变”的含义，这是从简单相关关系走向因果分析的第一步。
3. 但本章只是线性回归的初步演示。我们先学习如何建立模型、运行 OLS、阅读系数和 p 值；更严格的因果论证还需要进一步讨论遗漏变量、选择偏误、反向因果、工具变量、固定效应、自然实验等问题。
4. 因此，本章的目标是服务于因果分析，但方法还没有完全展开。

**本章知识点**

1. 多元线性回归：用多个解释变量解释一个连续因变量。
2. 回归公式：区分因变量、自变量、截距和误差项。
3. OLS 估计：用 `statsmodels` 建立并拟合普通最小二乘模型。
4. 回归结果解读：重点看系数、标准误、t 值、p 值和 $R^2$。
5. 稳健标准误：初步了解异方差稳健标准误的写法和作用。
6. 练习与拓展：用就业培训数据练习带有因果含义的多元回归，选学固定效应模型。


## 基本多元回归


**本章使用的数据**

本章使用的数据主要放在工作目录下的 `data` 文件夹中。读取时用相对路径，例如 `data/wage1.csv`。

- `data/wage1.csv`：伍德里奇《计量经济学导论》中的工资数据。一行是一名劳动者，包含工资、教育年数、工作经验、在当前工作的年限、性别等变量，用于演示多元线性回归。[下载](data/wage1.csv)
- `data/jtrain2.csv`：Lalonde 就业培训实验数据。一行是一名参加 National Supported Work Demonstration 项目的劳动者，包含是否被分配到就业培训、培训前收入、培训后收入、年龄、教育年数等变量，用于本章新版作业。`train` 表示是否被分配到培训，`re78` 表示 1978 年真实收入，单位为千美元。该数据的培训分配接近随机实验，因此比一般观察数据更适合讨论因果解释。[下载](data/jtrain2.csv)，[变量说明](data/jtrain2_des.txt)
- `data/gpa1.csv`：学生成绩数据。一行是一名学生，包含大学第一年 GPA、高中 GPA、ACT 成绩等变量。该数据保留为旧版作业示例，当前已标记为废弃。[下载](data/gpa1.csv)
- `grunfeld`：`statsmodels` 内置的企业投资面板数据，通过 `statsmodels.datasets.grunfeld` 读取，不需要单独下载，用于选学固定效应示例。

::: {.callout-important}
### 注意
1. 本章默认本地数据文件放在当前章节 notebook 同级目录下的 `data` 文件夹中。
2. 出现 `FileNotFoundError`，通常是工作目录不对，或者数据文件没有放在 `data` 文件夹中。
3. 固定效应部分属于选学内容，重点是看懂面板数据模型的基本形式和代码写法。
:::



### 分析目标和数据


这个案例来自伍德里奇的《计量经济学导论》。

本案关心的是一个因果问题：**教育水平是否会影响工资？即：每多读一年书，工资是否会提高？**

要注意，提出因果问题并不等于已经完成因果论证。仅仅做一个横截面 OLS 回归，还不能自动证明教育对工资的因果效应。本章先用这个例子演示回归模型的建立、OLS 的运行和回归系数的解释；后续更严格的因果识别问题再继续展开。

数据采用《导论》中的数据`Wage1`。

In [25]:
# 加载并查看wage1.csv数据集
import pandas as pd

# 加载数据
wage1_data = pd.read_csv("data/wage1.csv")

# 查看数据集的前几行和列名
wage1_data.iloc[:5, :10]  # 显示前5行和前10列

,wage,educ,exper,tenure,nonwhite,female,married,numdep,smsa,northcen
0,3.10,11,2,0,0,1,0,2,1,0
1,3.24,12,22,2,0,1,1,3,1,0
2,3.00,11,2,0,0,0,0,2,0,0
3,6.00,8,44,28,0,0,1,0,1,0
4,5.30,12,7,2,0,0,1,1,0,0


In [26]:
wage1_data.columns # 显示所有列名

Index(['wage', 'educ', 'exper', 'tenure', 'nonwhite', 'female', 'married',
       'numdep', 'smsa', 'northcen', 'south', 'west', 'construc', 'ndurman',
       'trcommpu', 'trade', 'services', 'profserv', 'profocc', 'clerocc',
       'servocc', 'lwage', 'expersq', 'tenursq'],
      dtype='object')

### 回归公式

我们的目标是讨论教育水平对工资的因果影响。为了比简单相关关系更进一步，可以建立以下多元回归模型：

$$
\text{wage} = \beta_0 + \delta_0 \cdot \text{female} + \beta_1 \cdot \text{educ} + \beta_2 \cdot \text{exper} + \beta_3 \cdot \text{tenure} + u
$$

其中：

- $\text{wage}$ 是工资。
- $\text{female}$ 是一个虚拟变量，女性为1，男性为0。
- $\text{educ}$, $\text{exper}$, 和 $\text{tenure}$ 分别代表教育年数、工作经验和在当前工作的年限。
- $u$ 是误差项。

从理论上看，这类线性回归模型通常可以写成一般形式：
$$
y = \beta_0 + \beta_1 x_1 + \cdots + \beta_k x_k + u,
$$
其中 $E(u\mid x_1,\dots,x_k)=0$ 是一个核心假设，它表示在给定解释变量之后，误差项的条件期望为零。普通最小二乘法（OLS）就是通过让残差平方和尽量小来估计这些系数。

**教育水平对工资的影响体现在系数$\beta_1$中，这就是我们关心的目标。**

### 为什么要加入控制变量

在经济学中，回归通常不是为了停留在“相关关系”，而是为了讨论因果关系。加入控制变量，是从简单相关关系走向因果分析的第一步。

如果只看教育年数和工资的关系，我们得到的是两者的总体相关关系。但教育年数较高的人，可能同时有不同的工作经验、性别结构、行业选择、家庭背景、能力、地区机会等。如果这些因素也会影响工资，那么教育和工资之间的简单关系就可能混入了其他因素的影响。

多元回归的作用，就是在模型中加入一些可以观测到的重要因素作为控制变量。比如在本例中，我们加入：

- `female`：控制性别差异；
- `exper`：控制总工作经验；
- `tenure`：控制在当前工作的年限。

加入这些变量以后，教育系数 $\beta_1$ 的含义就发生了变化。它不再是“教育年数和工资的简单关系”，而是在 `female`、`exper` 和 `tenure` 保持不变的条件下，教育年数变化时工资的平均变化。

因此，控制变量让斜率有了“其他条件保持不变”的含义。这正是多元回归比简单相关分析更接近因果分析的地方。但这只是第一步。只有当重要的混杂因素都被恰当地控制，或者有更强的研究设计支持时，回归系数才更有资格被解释为因果效应。本章先从最基本的控制变量思想开始。

### 进行回归分析

在进行回归分析之前，我们需要准备数据和定义模型。这个过程包括设置自变量（解释变量）和因变量（被解释变量），以及确保数据格式适合进行回归分析。

#### 定义因变量和自变量

- **因变量** $y$: 因变量是我们想要预测或解释的变量。在本例中，因变量是 `wage`，即工资，我们希望了解其他变量如何影响工资。
  
- **自变量** $X$: 自变量是可能影响因变量的变量。在这个模型中，我们选择了四个自变量：`female`, `educ`, `exper`, 和 `tenure`。这些变量分别表示：
  
  - `female`: 表示性别，是一个二分类变量，女性为1，男性为0。
  - `educ`: 表示受教育年数，是一个连续变量。
  - `exper`: 表示总工作经验年数，是一个连续变量。
  - `tenure`: 表示在当前工作的年限，也是一个连续变量。

#### 构造自变量 $X$ 和因变量 $y$

在开始构建模型之前，我们需要从数据集中提取这些变量。使用Pandas库，我们可以从DataFrame中选择需要的列。同时，为了进行回归分析，我们需要向自变量矩阵 $X$ 中添加一个常数项，以便模型包含截距 $\beta_0$。


In [27]:
import statsmodels.api as sm

# 选择自变量和因变量
X = wage1_data[["female", "educ", "exper", "tenure"]]
y = wage1_data["wage"]

# 向自变量矩阵X添加常数项，以便包含截距
X = sm.add_constant(X)
X.head() # 注意，多了一列const

,const,female,educ,exper,tenure
0,1.0,1,11,2,0
1,1.0,1,12,22,2
2,1.0,0,11,2,0
3,1.0,0,8,44,28
4,1.0,0,12,7,2



#### 拟合回归模型

使用`statsmodels`库的`OLS`类，我们可以定义并拟合一个普通最小二乘回归模型。`OLS`类的第一个参数是因变量 $y$，第二个参数是自变量 $X$。在前面的模型假设下，OLS 给出的系数是通过最小化残差平方和得到的估计值，可以用来构造样本回归函数（或预测方程）。

In [28]:
# 创建OLS回归模型的实例并拟合数据
model = sm.OLS(y, X)
results = model.fit()

#### 查看回归结果

模型拟合后，我们可以通过调用 `summary()` 方法来查看详细的回归分析结果，包括每个变量的系数、标准误、t 值、p 值等统计指标。每个系数的 t 值和 p 值都对应于一个假设检验，通常是检验原假设 $H_0: \beta_j = 0$，用来判断该解释变量在统计上是否对因变量有影响。

In [29]:
# 输出回归结果摘要
print(results.summary())

                            OLS Regression Results                            
Dep. Variable:                   wage   R-squared:                       0.364
Model:                            OLS   Adj. R-squared:                  0.359
Method:                 Least Squares   F-statistic:                     74.40
Date:                Sun, 07 Jun 2026   Prob (F-statistic):           7.30e-50
Time:                        23:06:26   Log-Likelihood:                -1314.2
No. Observations:                 526   AIC:                             2638.
Df Residuals:                     521   BIC:                             2660.
Df Model:                           4                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         -1.5679      0.725     -2.164      0.0

### 稳健标准误

上面的回归表默认使用常规标准误。如果截面数据存在异方差，常规标准误可能不够可靠。一个常见做法是报告异方差稳健标准误。这里先不展开理论，只演示如何在 `statsmodels` 中取得 HC1 稳健标准误，并和常规标准误作比较。

In [30]:
# 比较常规标准误和HC1稳健标准误
robust_results = results.get_robustcov_results(cov_type="HC1")

robust_table = pd.DataFrame(
    {
        "coef": results.params,
        "常规标准误": results.bse,
        "HC1稳健标准误": robust_results.bse,
        "常规p值": results.pvalues,
        "HC1稳健p值": robust_results.pvalues,
    },
    index=results.params.index,
)

robust_table.round(3)

,coef,常规标准误,HC1稳健标准误,常规p值,HC1稳健p值
const,-1.568,0.725,0.826,0.031,0.058
female,-1.811,0.265,0.254,0.000,0.000
educ,0.572,0.049,0.061,0.000,0.000
exper,0.025,0.012,0.010,0.029,0.010
tenure,0.141,0.021,0.028,0.000,0.000


### 样本回归函数

可以得到样本回归函数（或者预测方程）：

$$
\widehat{wage} = -1.568 + (-1.811) \cdot female + 0.572 \cdot educ + 0.025 \cdot exper + 0.141 \cdot tenure
$$

这里给出的就是样本回归函数，用估计得到的系数来给出在不同特征组合下的工资预测值。与前面的“包含误差项 $u$ 的模型方程”不同，这里已经不再带误差项。

因为这是多元回归，严格说它不是二维图形中的一条“回归线”，而是包含多个解释变量的预测方程。

### 解释回归结果

从回归结果中，我们可以提取以下关键信息：

- **模型拟合情况**: $R^2$ 值为 0.364，说明模型解释了 36.4% 的因变量（wage）的方差。这表明模型有一定的解释能力，但在线性回归中，我们通常不会仅凭 $R^2$ 来判断模型优劣，更关心的是系数的含义及其显著性。
- **系数解释**:
  - **const**（截距）: $-1.568$。在常规标准误下，截距项的 p 值小于 0.05；但在 HC1 稳健标准误下，p 值约为 0.058，不再达到 5% 显著性水平。截距在形式上表示当所有解释变量为0时预测的工资值。在本例中，“教育年数=0、工作经验=0、在现单位年限=0、且为男性”的情形本身并不常见，因此截距更多是起到调整预测方程位置的技术作用，通常不作强经济解释。
  - **female**: $-1.811$，显著为负，表示在控制教育、总工作经验和在现单位年限之后，女性的工资平均比男性低约 \$1.81。这不是简单的男女平均工资差，而是在这些控制变量相同条件下的差异。
  - **educ**（教育年数）: $0.572$，显著为正。它是本例最关心的系数，因为它对应我们提出的因果问题：教育年数增加是否会提高工资。在这个简单回归中，如果暂时接受控制变量已经处理了主要混杂因素，那么可以按因果问题的语言理解为：在性别、总工作经验和当前工作年限相同的条件下，教育年数每增加一年，工资平均提高约 \$0.57。
  - **exper**（工作经验）: $0.025$，在 5% 水平下显著，表示在其他变量保持不变时，每增加一年总工作经验，工资平均增加约 \$0.03。
  - **tenure**（在当前工作的年限）: $0.141$，显著为正，表示在其他变量保持不变时，每增加一年在当前工作的年限，工资平均增加约 \$0.14。

- **模型诊断**:
  - **Durbin-Watson** 统计量为 1.794，接近2，未显示明显的一阶自相关迹象。不过本例是截面数据，Durbin-Watson 不是最核心的诊断指标。
  - **Omnibus** 和 **Jarque-Bera** 测试的 $p$ 值都接近0，表明残差不符合正态分布。在截面回归中，也常结合稳健标准误来处理异方差带来的标准误问题。

### 回归结果的经济学解释

在经济学研究中，分析者往往特别关注模型中的系数（$\beta_i$），因为这些系数对应我们关心的因果问题。教育系数 $0.572$ 的含义，不是简单地说“教育年数多的人工资更高”，而是：在性别、总工作经验和当前工作年限相同的条件下，教育年数每增加一年，工资平均提高约 0.57 美元。这种“其他条件保持不变”的解释，正是加入控制变量后的斜率含义。

但要注意，这个因果解释要成立，需要更强的论证：例如教育年数不能和遗漏在误差项中的能力、家庭背景、地区、职业选择等因素系统相关。本章先学习回归的基本操作和系数解读，后续再讨论如何更严格地支持这种因果解释。

#### 对系数的解释

1. **大小**：系数的大小告诉我们，当自变量增加一个单位时，因变量平均会增加或减少多少。

2. **方向**：系数的符号（正或负）表明变量作用方向。例如，如果某个变量的系数为正，则表明该变量增加时，被解释变量平均上升；如果系数为负，则平均下降。

3. **显著性**：统计显著性（通常通过p值来衡量）告诉我们系数是否在统计上显著，即我们观察到的关系是否有可能仅仅是由于抽样误差。p值较小（通常小于0.05）意味着我们有足够的证据认为该系数在总体中不为零。

### 多元回归作业：就业培训是否提高收入

这道作业使用 `data/jtrain2.csv`。数据来自 Lalonde 的就业培训研究，研究对象是参加 National Supported Work Demonstration 项目的劳动者。这个项目中，是否被分配到就业培训具有实验性质，因此本题比普通的成绩预测练习更适合讨论因果关系。

**研究问题：**

就业培训是否会提高劳动者之后的收入？

**主要变量：**

- `train`：是否被分配到就业培训，1 表示被分配到培训组，0 表示对照组。
- `re78`：1978 年真实收入，单位为千美元，是本题的因变量。
- `age`：年龄。
- `educ`：受教育年数。
- `black`：是否为黑人。
- `hisp`：是否为西班牙裔。
- `married`：是否已婚。
- `nodegree`：是否没有高中学历。
- `re74`、`re75`：1974 年和 1975 年真实收入，单位为千美元，表示培训前收入。

**本题用到的方法和函数：**

- `pd.read_csv()`：读取本地 CSV 文件。
- `DataFrame.describe()`：查看变量的描述性统计。
- `sm.add_constant()`：为自变量矩阵添加截距项。
- `sm.OLS(y, X).fit()`：估计普通最小二乘回归模型。
- `get_robustcov_results(cov_type="HC1")`：得到 HC1 异方差稳健标准误。

**任务：**

1. 读取数据，查看核心变量的描述性统计。
2. 估计不加控制变量的模型：

$$
re78_i = \beta_0 + \beta_1 train_i + u_i
$$

3. 估计加入控制变量的模型：

$$
re78_i = \beta_0 + \beta_1 train_i + \beta_2 age_i + \beta_3 educ_i + \beta_4 black_i + \beta_5 hisp_i + \beta_6 married_i + \beta_7 nodegree_i + \beta_8 re74_i + \beta_9 re75_i + u_i
$$

4. 比较两个模型中 `train` 的系数和显著性。
5. 解释 `train` 的系数含义：在什么条件下，它可以被理解为就业培训对收入的因果影响？

In [31]:
# 读取就业培训数据并查看前几行
jtrain2_data = pd.read_csv("data/jtrain2.csv")

jtrain2_data.head()

,train,age,educ,black,hisp,married,nodegree,mosinex,re74,re75,re78,unem74,unem75,unem78,lre74,lre75,lre78,agesq,mostrn
0,1,37,11,1,0,1,1,13,0.0,0.0,9.93005,1,1,0,0.0,0.0,2.295566,1369,13
1,1,22,9,0,1,0,1,13,0.0,0.0,3.59589,1,1,0,0.0,0.0,1.279792,484,13
2,1,30,12,1,0,0,0,13,0.0,0.0,24.90950,1,1,0,0.0,0.0,3.215249,900,13
3,1,27,11,1,0,0,1,13,0.0,0.0,7.50615,1,1,0,0.0,0.0,2.015723,729,13
4,1,33,8,1,0,0,1,13,0.0,0.0,0.28979,1,1,0,0.0,0.0,-1.238599,1089,13


In [32]:
# 查看作业核心变量的描述性统计
jtrain2_vars = ["train", "re78", "age", "educ", "black", "hisp", "married", "nodegree", "re74", "re75"]

jtrain2_data[jtrain2_vars].describe().round(2)  # round(2)保留2位小数，便于阅读

,train,re78,age,educ,black,hisp,married,nodegree,re74,re75
count,445.00,445.00,445.00,445.00,445.00,445.00,445.00,445.00,445.00,445.00
mean,0.42,5.30,25.37,10.20,0.83,0.09,0.17,0.78,2.10,1.38
std,0.49,6.63,7.10,1.79,0.37,0.28,0.37,0.41,5.36,3.15
min,0.00,0.00,17.00,3.00,0.00,0.00,0.00,0.00,0.00,0.00
25%,0.00,0.00,20.00,9.00,1.00,0.00,0.00,1.00,0.00,0.00
50%,0.00,3.70,24.00,10.00,1.00,0.00,0.00,1.00,0.00,0.00
75%,1.00,8.12,28.00,11.00,1.00,0.00,0.00,1.00,0.82,1.22
max,1.00,60.31,55.00,16.00,1.00,1.00,1.00,1.00,39.57,25.14


In [33]:
# 估计不加控制变量的就业培训回归
X_train_simple = jtrain2_data[["train"]]
X_train_simple = sm.add_constant(X_train_simple)  # 添加截距项
Y_train = jtrain2_data["re78"]

train_simple_results = sm.OLS(Y_train, X_train_simple).fit()
print(train_simple_results.summary())

                            OLS Regression Results                            
Dep. Variable:                   re78   R-squared:                       0.018
Model:                            OLS   Adj. R-squared:                  0.016
Method:                 Least Squares   F-statistic:                     8.039
Date:                Sun, 07 Jun 2026   Prob (F-statistic):            0.00479
Time:                        23:06:26   Log-Likelihood:                -1468.8
No. Observations:                 445   AIC:                             2942.
Df Residuals:                     443   BIC:                             2950.
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const          4.5548      0.408     11.162      0.0

In [34]:
# 估计加入控制变量后的就业培训回归
control_vars = ["train", "age", "educ", "black", "hisp", "married", "nodegree", "re74", "re75"]
X_train_control = jtrain2_data[control_vars]
X_train_control = sm.add_constant(X_train_control)  # 添加截距项

train_control_results = sm.OLS(Y_train, X_train_control).fit()
print(train_control_results.summary())

                            OLS Regression Results                            
Dep. Variable:                   re78   R-squared:                       0.055
Model:                            OLS   Adj. R-squared:                  0.035
Method:                 Least Squares   F-statistic:                     2.804
Date:                Sun, 07 Jun 2026   Prob (F-statistic):            0.00329
Time:                        23:06:26   Log-Likelihood:                -1460.2
No. Observations:                 445   AIC:                             2940.
Df Residuals:                     435   BIC:                             2981.
Df Model:                           9                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const          0.7851      3.375      0.233      0.8

In [35]:
# 比较两个模型中train系数的常规标准误和HC1稳健标准误
simple_robust = train_simple_results.get_robustcov_results(cov_type="HC1")
control_robust = train_control_results.get_robustcov_results(cov_type="HC1")

train_compare = pd.DataFrame(
    {
        "模型": ["不加控制变量", "加入控制变量"],
        "train系数": [train_simple_results.params["train"], train_control_results.params["train"]],
        "常规p值": [train_simple_results.pvalues["train"], train_control_results.pvalues["train"]],
        "HC1稳健p值": [simple_robust.pvalues[1], control_robust.pvalues[1]],
    }
)

train_compare.round(3)

,模型,train系数,常规p值,HC1稳健p值
0,不加控制变量,1.794,0.005,0.008
1,加入控制变量,1.676,0.009,0.014


参考解读：在不加控制变量的模型中，`train` 的系数约为 1.794；加入年龄、教育、族裔、婚姻、高中学历和培训前收入等控制变量后，`train` 的系数约为 1.676。由于 `re78` 的单位是千美元，这表示培训组在 1978 年的收入平均高出约 1.7 千美元。

这道题的重点不是只看 `train` 和 `re78` 是否相关，而是讨论就业培训对收入的因果影响。因为培训分配具有实验性质，培训组和对照组在原则上更具有可比性；加入控制变量以后，`train` 的系数可以理解为在年龄、教育、婚姻、族裔和培训前收入等条件保持不变时，培训组与对照组的平均收入差异。因此，这个作业比 GPA 作业更适合训练因果解释。



### 多元回归练习题（废弃）

::: {.callout-warning}
这一版作业保留作为旧示例，不再作为本章推荐作业。它更适合练习条件相关关系和预测，不够适合作为因果关系训练。新版作业请使用上面的就业培训数据。
:::

在一项研究中，研究人员收集了一批学生的高中成绩（hsGPA）、大学入学考试成绩（ACT）以及他们在大学第一年的成绩（colGPA）。该研究目的是探索高中GPA和ACT成绩对大学第一年GPA的影响。

**数据集包含以下变量：**

- `hsGPA`：高中GPA
- `ACT`：大学入学考试成绩
- `colGPA`：大学第一年GPA

**研究问题：**
高中GPA和ACT成绩如何影响大学第一年的GPA？

**回归模型：**
使用以下回归模型分析数据：
$$
\text{colGPA} = \beta_0 + \beta_1 \times \text{hsGPA} + \beta_2 \times \text{ACT} + u
$$

其中：

- $\beta_0$ 是截距
- $\beta_1$ 是高中GPA对大学GPA的影响。
- $\beta_2$ 是ACT成绩对大学GPA的影响。
- $u$ 是误差项。

在这个模型中，可以把系数理解为在控制了另一个成绩之后，各自对大学第一年GPA的边际影响。需要注意的是，如果还有其他重要但未观测的因素同时影响 hsGPA、ACT 和 colGPA，这些遗漏变量可能会使系数偏离理想的因果效应，因此在解释时要结合实际背景进行判断。

**任务：**

1. 利用提供的数据，估计上述回归模型。
2. 解释高中GPA和ACT成绩的系数，包括意义和显著性。
3. 思考题：讨论模型的适用性和限制。



In [36]:
# 读取数据
gpa1_data = pd.read_csv("data/gpa1.csv")

# 显示列名
gpa1_data.columns

Index(['age', 'soph', 'junior', 'senior', 'senior5', 'male', 'campus',
       'business', 'engineer', 'colGPA', 'hsGPA', 'ACT', 'job19', 'job20',
       'drive', 'bike', 'walk', 'voluntr', 'PC', 'greek', 'car', 'siblings',
       'bgfriend', 'clubs', 'skipped', 'alcohol', 'gradMI', 'fathcoll',
       'mothcoll'],
      dtype='object')

下面给出这道练习的参考实现。先构造自变量矩阵并添加常数项，再用 `OLS` 拟合模型。

In [37]:
# 构造并估计GPA回归模型
X_gpa = gpa1_data[["hsGPA", "ACT"]]
X_gpa = sm.add_constant(X_gpa)
y_gpa = gpa1_data["colGPA"]

gpa_results = sm.OLS(y_gpa, X_gpa).fit()
print(gpa_results.summary())

                            OLS Regression Results                            
Dep. Variable:                 colGPA   R-squared:                       0.176
Model:                            OLS   Adj. R-squared:                  0.164
Method:                 Least Squares   F-statistic:                     14.78
Date:                Sun, 07 Jun 2026   Prob (F-statistic):           1.53e-06
Time:                        23:06:26   Log-Likelihood:                -46.573
No. Observations:                 141   AIC:                             99.15
Df Residuals:                     138   BIC:                             108.0
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const          1.2863      0.341      3.774      0.0

参考解读：`hsGPA` 的系数约为 0.454，且 p 值很小，说明在控制 ACT 之后，高中 GPA 每提高 1 分，大学第一年 GPA 平均提高约 0.454 分。`ACT` 的系数约为 0.009，p 值约为 0.383，在常用显著性水平下不显著。如果要把这些系数解释为因果效应，还需要进一步说明是否存在遗漏变量、选择偏误或反向因果等问题。这里先练习控制变量条件下的系数解读。

## 选学：固定效应

更进一步，我们可以使用 `linearmodels`来完成基准回归，并且加入固定效应。本节属于面板数据回归的预告，重点是看懂基本形式和代码写法。

在面板数据中，同一企业在不同年份反复出现，每个企业往往有一些不随时间变化但又影响因变量的特质，例如管理水平、企业文化等。如果这些特质与解释变量有关，简单的回归可能会有偏。固定效应模型通过为每个个体（以及每个年份）引入一个单独的截距，把这种“时间不变但个体不同”的因素吸收掉。

固定效应可以控制企业不随时间变化的特征，以及年份共同冲击，因此比完全不控制这些因素的模型更接近我们想要的因果比较。但它仍然不能自动解决所有问题：如果存在随时间变化、同时影响投资和企业价值的遗漏因素，或者存在反向因果，系数仍然不能简单视为严格因果效应。

如果导入 `linearmodels` 时出错，可以在Jupyter Notebook的python单元格中执行如下代码，注意开头有个感叹号：

`!pip install -i https://pypi.tuna.tsinghua.edu.cn/simple linearmodels`

### 数据准备

这里使用来自statsmodels的数据集 [grunfeld](https://www.statsmodels.org/stable/datasets/generated/grunfeld.html) ：“The Determinants of Corporate Investment”

这里我们试图使用企业的价值和资本来解释企业的投资行为。

In [38]:
from statsmodels.datasets import grunfeld

data = grunfeld.load_pandas().data
data.head()

,invest,value,capital,firm,year
0,317.6,3078.5,2.8,General Motors,1935.0
1,391.8,4661.7,52.6,General Motors,1936.0
2,410.6,5387.1,156.9,General Motors,1937.0
3,257.7,2792.2,209.2,General Motors,1938.0
4,330.8,4313.2,203.4,General Motors,1939.0


变量的定义如下：

- invest  - Gross investment in 1947 dollars
- value   - Market value as of Dec. 31 in 1947 dollars
- capital - Stock of plant and equipment in 1947 dollars
- firm    - General Motors, US Steel, General Electric, Chrysler, Atlantic Refining, IBM, Union Oil, Westinghouse, Goodyear, Diamond Match, American Steel
- year    - 1935 - 1954

首先把表示个体和时间的列（准备要控制的固定效应）都转为index

In [39]:
data = data.set_index(["firm", "year"])
data.head()


invest   value  capital
firm           year                           
General Motors 1935.0   317.6  3078.5      2.8
               1936.0   391.8  4661.7     52.6
               1937.0   410.6  5387.1    156.9
               1938.0   257.7  2792.2    209.2
               1939.0   330.8  4313.2    203.4

### 进行回归

考虑企业投资的决定，我们设置2个回归模型：

1. 不带有固定效应:
$$
invest_{it} = \beta_0 + \beta_1 value_{it} + \beta_2 capital_{it} + \varepsilon_{it}
$$

2. 带有个体和时间固定效应，其中$\nu_i$、$\lambda_t$分别代表个体和时间固定效应：
$$
invest_{it} = \beta_0 + \beta_1 value_{it} + \beta_2 capital_{it} + \nu_i + \lambda_t +  \varepsilon_{it}
$$

这里 $\nu_i$ 可以理解为企业 $i$ 的“固定特征”，在所有年份都相同；$\lambda_t$ 是年份 $t$ 的共同冲击，对所有企业都相同。例如，宏观经济环境的变化就可以通过年份固定效应来近似刻画。

要控制固定效应，只要在公式中添加 `EntityEffects`或`TimeEffects` 即可。拟合两个模型：

In [40]:
from linearmodels import PanelOLS

# 无固定效应
m1 = PanelOLS.from_formula(
    "invest ~ 1 + value + capital", data=data
).fit()


# 有固定效应
m2 = PanelOLS.from_formula(
    "invest ~ 1 + value + capital + EntityEffects + TimeEffects", data=data
).fit()


# 有固定效应，聚类到个体
m3 = PanelOLS.from_formula(
    "invest ~ 1 + value + capital + EntityEffects + TimeEffects", data=data
).fit(cov_type="clustered", cluster_entity=True)

比较结果：

In [41]:
from linearmodels.panel.results import compare

print(compare(results = {'OLS':m1,"FE":m2, 'FE-cluster':m3}))

                          Model Comparison                         
                                   OLS             FE    FE-cluster
-------------------------------------------------------------------
Dep. Variable                   invest         invest        invest
Estimator                     PanelOLS       PanelOLS      PanelOLS
No. Observations                   220            220           220
Cov. Est.                   Unadjusted     Unadjusted     Clustered
R-squared                       0.8179         0.7253        0.7253
R-Squared (Within)              0.7357         0.7566        0.7566
R-Squared (Between)             0.8426         0.7944        0.7944
R-Squared (Overall)             0.8179         0.7856        0.7856
F-statistic                     487.28         248.15        248.15
P-value (F-stat)                0.0000         0.0000        0.0000
=====================     ============   ============   ===========
Intercept                      -38.410        -7

注意到

1. 2个“FE”列的最下方，显示这2个模型已经控制了个体和时间固定效应。
2. Cov. Est.行， 显示FE-Cluster模型使用了聚类标准误。
